Add Silent Payments Support

macgyver13 opened the pull request:
Finally consider this PR ready for review and feedback to determine next steps.

## TODOs:
- ~~verify silent payments labeled address handling~~
- ~~verify multi sig silent payment workflows~~
- ~~address edge cases / attack scenarios~~
- (on hold) test on real hardware - evaluate memory and performance
- (partial) proper UI/UX for Silent Payments - **seeking feedback**

## Screenshots
### Updated UX with - silent payment address - appended to - to address -
<img width="384" height="300" alt="SHIFT SYM CAPS" src="https://github.com/user-attachments/assets/0679c328-82c0-467d-bdb8-9be278db54e2" />
<img width="398" height="300" alt="SHIFT SYM CAPS" src="https://github.com/user-attachments/assets/3a13fbdf-0f10-41e5-a434-af832a96b06f" />
<img width="388" height="296" alt="SHIFT SYM CAPS" src="https://github.com/user-attachments/assets/7bf67a5f-4848-4330-8a75-ef0a2017f662" />

### Contribute Shares Prompt 
<img width="367" height="283" alt="0F056943EDGE" src="https://github.com/user-attachments/assets/921285d5-6e78-4106-8232-39e079a907b1" />

macgyver13 committed:
BIP-352: implement Silent Payments cryptographic primitives

Add bip352 crypto primitives / helpers
Add bip352 tagged hashes

macgyver13 committed:
BIP-374: implement DLEQ proof generation and verification

Add bip374 tagged hashes
Add dleq bip374 reference generate_dleq_proof, verify_dleq_proof functions

macgyver13 committed:
BIP-375: add test vectors

Use devtest/verify_sp_outputs.py for simulator testing 
Add testing/test_bip375_vectors.py and testing/bip375_test_vectors.json
- Correct input eligibility ecdh share assumptions
Add devtest/unit_silentpayments.py to perform unit testing via simulator
Mirror psbt fields in testing/psbt.py

macgyver13 committed:
BIP-375: implement Silent Payments PSBT signing workflows

Integrate preview silent payment address as output in auth:interact
Integrate silent payments in signing workflow
Add sp_hrp property to chains.py
Add silent payments fields to psbt.py

macgyver13 commented on shared/bech32.py:
```diff
@@ -0,0 +1,135 @@
+# from <https://github.com/sipa/bech32/blob/master/ref/python/segwit_addr.py>
```
Good catch! Thank you for the feedback.

Unfortunately, I could not use any existing encoding functions in ngu so I followed the scheme of adding nip19 encoding to add bip352_encode.  If you have a better idea please let me know.

scgbckbone replied:
yeah, I completely forgot we only expose segwit_{encode/decode} thru `ngu.codecs` - sorry for unreliable info.

Your approach, adding it to the libngu, is correct.

macgyver13 committed:
BIP-376: handle spending silent payments outputs

Add sp_spend support to _derive_input_privkey
Validate bip376 spend key derivation

nymius commented on shared/silentpayments.py:
```diff
@@ -125,6 +125,45 @@ def compute_shared_secret_tweak(ecdh_share_bytes, k):
+    BIP-352 formula for spending: d_k = d_spend + t_k (mod n)
+    where:
+    - d_spend is the base spend private key
+    - t_k is the shared secret tweak from PSBT_IN_SP_TWEAK
```
The above line is stating correctly `d_spend` is "the base spend private key", but "`t_k` is the shared secret tweak" is ruling out the `label` tweak (both hashes have different tags). I guess you were trying to refer to it too, but I would add it explicitly.

macgyver13 committed:
nit: mark bitcoind dependent sign tests

macgyver13 committed:
SP: add Silent Payments integration tests

Create end-to-end tests for SP in testing/test_silentpayments.py
- Expose foreign_mk for partial ownership coverage test
Integrate SP spend into psbt functions: consider_inputs, determine_my_signing_key, sign_it
- Synthesize parsed_subpaths from PSBT_IN_SPEND_BIP32_DERIVATION
- Set input.sp_idxs to placeholder to synthesize input is_owned
- Implement SP input signing
Refactor _compute_and_store_ecdh_shares and _derive_input_privkey to leverage input.sp_idxs

macgyver13 committed:
fix: (psbt v2) structural changes

Only store key_data for short_values - remove key_type
Only serialize PSBT_OUT_SCRIPT if self.script has a value
Adjust v2 script assert to not required if sp_v0_info is present

macgyver13 committed:
SP: add SilentPaymentsMixin

Extend psbt handling with silent payment specific functions
Incorporate validation functions from bip375 test vectors
- psbt_structure, input_eligibility, ecdh_coverage
Add private key derivation helper functions

macgyver13 committed:
fix: (usb_test) Unify local and globals dict to single namespace

Provides silent payments testing framework with access to MockPSBT

macgyver13 committed:
SP: introduce partial share contribution workflow

Add test to validate 'Contribute Shares?' UX
Simplify SP preview and signing code paths

macgyver13 committed:
SP: introduce UI change output validation

Add silent payments label change tests
- Detect presence or absence of 'Change back:' in UI story
Add multi-signer incomplete coverage scenario tests

macgyver13 committed:
BIP-352: add test vectors

Add testing/test_bip352_vectors.py and testing/bip352_test_vectors.json 
- Verify SP sending and receiving test vectors

macgyver13 committed:
SP: add docs/silentpayments.md

macgyver13 committed:
SP: implement bip352 generic format export

Include key-value pairs: spscan, deriv, name, xfp, key_exp in bip352 dictionary
Add bip352 test block to test_export_coldcard
Support 127 character limit for silent payments v0

macgyver13 committed:
SP: add sp spend derivation path tests

Verify SP spend path was derived from m/352h
Add signature verification for taproot inputs and SP outputs

macgyver13 committed:
SP: add more integration tests

Mixed inputs SP spend and P2TR
Reject non-SIGHASH_ALL
Reject psbt v0 send to SP output

macgyver13 commented:
Summary of changes since March update:
- Added bip352_test_vectors.json and test_bip352_vectors.py
- Added BIP376 spend path validation
- Clear sensitive material in Silent Payments workflows
- Added doc/silentpayments.md
- Added bip352 generic format export
- Total integration tests in test_silentpayments.py:
  - test_sp_signing_story
  - test_sp_p2wpkh_input
  - test_sp_p2tr_input
  - test_sp_mixed_inputs
  - test_sp_mixed_sp_spend_and_regular_input
  - test_sp_three_outputs_same_scan_key
  - test_sp_two_outputs_different_scan_keys
  - test_sp_mixed_output_types
  - test_sp_all_owned_multi_input
  - test_sp_partial_owned_coverage_complete
  - test_sp_partial_owned_coverage_incomplete
  - test_sp_partial_owned_coverage_incomplete_refused
  - test_sp_spend_silent_payment_output
  - test_sp_spend_silent_payment_output_with_taproot
  - test_sp_spend_wrong_tweak_rejected
  - test_sp_spend_wrong_purpose_rejected
  - test_sp_spend_wrong_coin_type_rejected
  - test_exit_gracefully_on_sp_validation_failure
  - test_sp_non_sighash_all_rejected
  - test_sp_psbt_v0_rejected
  - test_sp_spend_to_labeled_change_address

## Testing

### Targeted tests
```
cd testing
pytest test_bip352_vectors.py
pytest test_bip375_vectors.py
pytest test_silentpayments.py
pytest test_export.py::test_export_coldcard
```
### General tests
DEFAULT_PYTEST_MARKS = "not onetime and not veryslow and not manual **and not bitcoind**"
python run_sim_tests.py --multiproc --num-proc 6

macgyver13 committed:
SP (refactor): test hygiene

Remove MockPSBT from verify_sp_outputs
Test test_sp_spend_to_labeled_change_address should sign transaction
Replace _serialize_psbt with read_psbt
Replace single use compute_foreign_share with generate_dleq_proof
Remove unnecessary int-byte conversions in test_bip352_vectors.py

macgyver13 committed:
SP: improve sp spend without sp outputs and change detection

Convert validate_bip376_spend to validate_silent_payment_inputs
Use accounts to validate sp spend material and mark output.is_change
Introduce reusable sp_derive_path function
Fix change detection issue raised by previous test hygiene commit
- Signing was restored for test_sp_spend_labeled_change_detected
Fix psbt exception in consider_outputs with sp change
- Add apply_label_to_spend_key primitive
- Match spend public key to labeled spend public key for change

New tests:
sp spend to non sp output
sp spend rejection tests - unknown account, incorrect sp spend derivation path
sp change output with foreign keys is allowed but displays warning
Changed test:
sp spend to label change - fix label spend key derivation, use sp input

macgyver13 committed:
fix (SP): apply sp output sort correctly when scan key is the same for multiple outputs

Per BIP-375: sort by spend key then output index in ascending order for each scan key group to determine ordering of the k value
- Fix compute_silent_payment_output_scripts to match spec 
- Update bip375_test_vectors.json to version 1.1.1

macgyver13 commented:
Commit `106990f` fixes silent payment output ordering when multiple outputs share the same scan key (labeled  addresses of one recipient).

macgyver13 committed:
feat (SP): hide export finalize txn option for silent payment outputs

Discourage broadcasting silent payments directly from the device. The wallet
coordinator should be responsible for verifying the PSBT and finalizing for 
broadcast.
- Users can restore the default ColdCard behavior by using the Danger Zone -> 
SP Final Export setting.
- USB API was not modified.

macgyver13 commented:
Commit e08bd9c modifies the default silent payment signing flow. The wallet coordinator should be responsible for verifying the PSBT and finalizing for broadcast when silent payments are involved.

scgbckbone commented:
am I the one blocking this with HW tests ? 

Please, open libngu changes as separate libngu PR.

macgyver13 commented:
> am I the one blocking this with HW tests ?

I consider this PR stable and ready for critical review. Confirming this PR on real hardware would be great!

> Please, open libngu changes as separate libngu PR.

I'll get on that. Thank you!

macgyver13 committed:
bump submodule version

external/ckcc-protocol
- Add bip375 and bip376 constants

external/libngu
- Expose ec_pubkey_combine
- Expose ec_pubkey_tweak_mul
- Add bip352_encode

update libngu

macgyver13 commented:
Last force push makes the following changes:
- update libngu submodule version to match PR
  - reduce bech32 decode character limit to match submodule for memory efficiency
- remove deadcode in dleq.py
- match bip v1.1.1 bip375_test_vectors.json
- add delay to stabilize test_sp_export* tests

scgbckbone reviewed (commented):
```
diff --git a/shared/manifest.py b/shared/manifest.py
index 6d4b087a..3aaed79a 100644
--- a/shared/manifest.py
+++ b/shared/manifest.py
@@ -17,6 +17,7 @@ freeze_as_mpy('', [
        'desc_utils.py',
        'dev_helper.py',
        'display.py',
+       'dleq.py',
        'drv_entro.py',
        'exceptions.py',
        'export.py',
@@ -52,6 +53,7 @@ freeze_as_mpy('', [
        'selftest.py',
        'serializations.py',
        'sffile.py',
+       'silentpayments.py',
        'ssd1306.py',
        'stash.py',
        'tapsigner.py',
```

Above patch fixes both Mk/Q HW builds. I didn't test the functionality in any way. Upgrade/downgrade works so you shouldn't stay locked if you have some issue in actual feature. 

Any new file that you add to shared needs to go to (specific) manifest(s). Otherwise it will not get bundled into .dfu and you fail with import errors before PIN entry. Simulator works without it and just inject those files.

macgyver13 committed:
fix (SP): add dleq.py and silentpayments.py to manifest

If shared files are not included in manifest bundling fails 
and will brick the device before PIN entry.
